### ISSUE TITLE
Odd negotiated IDs not working (gets shifted when setting description)

### ISSUE THREAD

**Issue by Faless** [paullouisageneau/libdatachannel#607]
Hi, first of all, thank you for making this great library :).
I've recently been working on switching the official [Godot WebRTC plugin](https://github.com/godotengine/webrtc-native) from the google library to libdatachannel, and I noticed one of our demo was failing.

Upon some digging I think I found an issue in the library causing channels created with `negotiated = true` and `id = 1` (or any other odd number) to fail:

See the failing test here: https://github.com/Faless/libdatachannel/blob/tests/negotiated/test/negotiated.cpp#L69

I believe this is caused by the peerconnection [shifting odd stream ids](https://github.com/paullouisageneau/libdatachannel/blob/master/src/impl/peerconnection.cpp#L293) ([even if they are negotiated](https://github.com/paullouisageneau/libdatachannel/blob/master/src/impl/peerconnection.cpp#L659)).
So, I tried a "dirty patch": https://github.com/Faless/libdatachannel/commit/5662720e04473bf34e449cf696ef3d8ed2ddd6b7

```diff
diff --git a/src/impl/datachannel.cpp b/src/impl/datachannel.cpp
index 622ad09a..0159afd4 100644
--- a/src/impl/datachannel.cpp
+++ b/src/impl/datachannel.cpp
@@ -85,7 +85,7 @@ bool DataChannel::IsOpenMessage(message_ptr message) {
 
 DataChannel::DataChannel(weak_ptr<PeerConnection> pc, uint16_t stream, string label,
                          string protocol, Reliability reliability)
-    : mPeerConnection(pc), mStream(stream), mLabel(std::move(label)),
+    : mPeerConnection(pc), mStream(stream), mNegotiated(true), mLabel(std::move(label)),
       mProtocol(std::move(protocol)),
       mReliability(std::make_shared<Reliability>(std::move(reliability))),
       mRecvQueue(RECV_QUEUE_LIMIT, message_size_func) {}
@@ -183,7 +183,7 @@ size_t DataChannel::maxMessageSize() const {
 
 void DataChannel::shiftStream() {
 	std::shared_lock lock(mMutex);
-	if (mStream % 2 == 1)
+	if (!mNegotiated && mStream % 2 == 1)
 		mStream -= 1;
 }
 
@@ -263,7 +263,9 @@ void DataChannel::incoming(message_ptr message) {
 
 NegotiatedDataChannel::NegotiatedDataChannel(weak_ptr<PeerConnection> pc, uint16_t stream,
                                              string label, string protocol, Reliability reliability)
-    : DataChannel(pc, stream, std::move(label), std::move(protocol), std::move(reliability)) {}
+    : DataChannel(pc, stream, std::move(label), std::move(protocol), std::move(reliability)) {
+	mNegotiated = false;
+}
 
 NegotiatedDataChannel::~NegotiatedDataChannel() {}
 
diff --git a/src/impl/datachannel.hpp b/src/impl/datachannel.hpp
index 5d9fe5a7..66c7ada0 100644
--- a/src/impl/datachannel.hpp
+++ b/src/impl/datachannel.hpp
@@ -69,6 +69,7 @@ struct DataChannel : Channel, std::enable_shared_from_this<DataChannel> {
 	weak_ptr<SctpTransport> mSctpTransport;
 
 	uint16_t mStream;
+	bool mNegotiated;
 	string mLabel;
 	string mProtocol;
 	shared_ptr<Reliability> mReliability;
```

Beside the name being a bit confusing (NegotiatedDataChannal should actually be OutgoingDataChannel I think) this mostly work, but there are still issue when mixing `negotiated = true` and `negotiated = false` data channels because the stream ID is always assigned before creating the session description, instead of doing the one that are `negotiated = false` during session description creation.
I.e.:
```js
var pc = new RTCPeerConnection();
pc.createDataChannel("test");
pc.createDataChannel("test2", {"negotiated": true, "id": 0})
```
is valid in browsers, because the stream ID of "test" is assigned when creating the session description and thus will not collide with the `id = 0` of "test2".

I hope I didn't miss anything, I think we should work in the direction of replacing `shiftDataChannels` with some `assignDataChannelsIDs`, let me know if it makes sense to you.

PS: Thanks again for making this library, I'm really looking forward to finish the porting of the Godot plugin.

**Comment by paullouisageneau** [comment]
I merged the fix in https://github.com/paullouisageneau/libdatachannel/pull/608 and renamed `NegotiatedDataChannel` to `IncomingDataChannel` in https://github.com/paullouisageneau/libdatachannel/pull/609.

I'll release the fix in a version 0.17.2 soon. Adapting the API with late stream id assignment like in browsers may be done later and released in version 0.18 as it'll break compatibility.

**Comment by paullouisageneau** [comment]
I'm closing this issue as the bug is fixed, and open a new one for changing the stream id assignment behavior.

**Comment by paullouisageneau** [comment]
Released in v0.17.2
